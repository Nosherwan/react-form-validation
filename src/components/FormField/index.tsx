import React, { Component, ChangeEvent } from 'react';
import debounce from 'lodash.debounce';

const styles = require('./styles.css');

export interface IFormField {
	label: string
	name: any
	value: string
	innerRef: any
	error: string
	changeHandler: (e: ChangeEvent<any>) => void
}

export class FormField extends Component<IFormField, any> {

	changeHandler: (e: ChangeEvent<any>) => void

	constructor(props: any) {
		super(props);
		this.changeHandler = debounce(this._changeHandler,200)
	}

	_typeHandler = (e: ChangeEvent<any>): void => {
		e.persist();
		this.changeHandler(e);
	}

	_changeHandler = (e: ChangeEvent<any>): void => {
		this.props.changeHandler(e);
	};

	render() {
		const { name, label, value, error, innerRef } = this.props;
		return (
			<div className={styles.field_container}>
				<label>
					{label}
				</label>
				<input
					ref={innerRef}
					defaultValue={value}
					name={name}
					className={error ? styles.invalid_field : styles.field}
					type='text'
					onChange={this._typeHandler}
				>
				</input>
				{error && (<div>{error}</div>)}
			</div>
		);
	}
}
