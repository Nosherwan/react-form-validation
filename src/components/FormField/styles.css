.field_container *{
  padding: 4px;
}

.field {
  border: 2px solid lightblue;
  border-radius: .25em;
}

.invalid_field {
  border: 2px solid red;
  border-radius: .25em;
}

.field_container {
  display: flex;
  flex-flow: column wrap;
  justify-items: center;
  justify-content: center;
  align-content: space-between;
  padding: 4px; 
}


.field_container input[type=text] {
  /* padding: 2px; */
}
