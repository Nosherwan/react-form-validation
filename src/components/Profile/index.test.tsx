import * as React from 'react';
import { shallow } from 'enzyme';
import { Profile, IProfile } from './index';
import { FormField } from '../FormField/index';

function getProfileProps(): IProfile {
  return {
    profile: {get: () => ''},
    getUser: () => true,
    updateUser: () => true,
  };
}

describe('Profile', () => {

  it('should render correctly', () => {
    const props = getProfileProps();
    const component = shallow(<Profile {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should have 3 FormField child components', () => {
    const props = getProfileProps();
    const component = shallow(<Profile {...props} />);
    expect(component.find(FormField)).toHaveLength(3);
  });

});
