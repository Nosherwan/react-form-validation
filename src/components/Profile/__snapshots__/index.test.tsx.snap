// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should render correctly 1`] = `
<div>
  <h2>
    Change Profile 
  </h2>
  <form
    className="profile_form"
    onSubmit={[Function]}
  >
    <FormField
      changeHandler={[Function]}
      error=""
      innerRef={[Function]}
      label="First Name"
      name="firstName"
      value=""
    />
    <FormField
      changeHandler={[Function]}
      error=""
      innerRef={[Function]}
      label="Last Name"
      name="lastName"
      value=""
    />
    <FormField
      changeHandler={[Function]}
      error=""
      innerRef={[Function]}
      label="Email"
      name="email"
      value=""
    />
    <div>
      <label>
        Select Status
      </label>
      <input
        checked=""
        name="available"
        onChange={[Function]}
        type="checkbox"
      />
    </div>
    <button
      className="disabled_submit_button"
      type="submit"
    >
       Submit 
    </button>
  </form>
</div>
`;
